third/cyrus-imapd/man/cyr_expire.8
.\" -*- nroff -*-
.TH CYR_EXPIRE 8 "Project Cyrus" CMU
.\" 
.\" Copyright (c) 1998-2003 Carnegie Mellon University.  All rights reserved.
.\"
.\" Redistribution and use in source and binary forms, with or without
.\" modification, are permitted provided that the following conditions
.\" are met:
.\"
.\" 1. Redistributions of source code must retain the above copyright
.\"    notice, this list of conditions and the following disclaimer. 
.\"
.\" 2. Redistributions in binary form must reproduce the above copyright
.\"    notice, this list of conditions and the following disclaimer in
.\"    the documentation and/or other materials provided with the
.\"    distribution.
.\"
.\" 3. The name "Carnegie Mellon University" must not be used to
.\"    endorse or promote products derived from this software without
.\"    prior written permission. For permission or any other legal
.\"    details, please contact  
.\"      Office of Technology Transfer
.\"      Carnegie Mellon University
.\"      5000 Forbes Avenue
.\"      Pittsburgh, PA  15213-3890
.\"      (412) 268-4387, fax: (412) 268-7395
.\"      tech-transfer@andrew.cmu.edu
.\"
.\" 4. Redistributions of any form whatsoever must retain the following
.\"    acknowledgment:
.\"    "This product includes software developed by Computing Services
.\"     at Carnegie Mellon University (http://www.cmu.edu/computing/)."
.\"
.\" CARNEGIE MELLON UNIVERSITY DISCLAIMS ALL WARRANTIES WITH REGARD TO
.\" THIS SOFTWARE, INCLUDING ALL IMPLIED WARRANTIES OF MERCHANTABILITY
.\" AND FITNESS, IN NO EVENT SHALL CARNEGIE MELLON UNIVERSITY BE LIABLE
.\" FOR ANY SPECIAL, INDIRECT OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
.\" WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN
.\" AN ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING
.\" OUT OF OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
.\" 
.\" $Id: cyr_expire.8,v 1.1.1.1 2004-02-23 22:54:13 rbasch Exp $
.SH NAME
cyr_expire \- expire messages and duplicate delivery database entries
.SH SYNOPSIS
.B cyr_expire
[
.B \-C
.I config-file
]
.BI \-E " days"
[
.B \-v
]
.SH DESCRIPTION
.I Cyr_expire
is used to expire messages and duplicate delivery database entries.
The expiration of messages is controlled by the
\fB/vendor/cmu/cyrus-imapd/expire\fR mailbox annotation which
specifies the age (in days) of messages in the given mailbox that
should be deleted.  Any duplicate delivery database entries which
correspond to the mailbox are also deleted at the same frequency.
.br
.sp
The value of the \fB/vendor/cmu/cyrus-imapd/expire\fR annotation is
inherited by all children of the given mailbox, so an entire mailbox
tree can be expired by seting a single annotation on the root of that
tree.  If a mailbox does not have a
\fB/vendor/cmu/cyrus-imapd/expire\fR annotation set on it (or does not
inherit one), then no messages are expired from the mailbox.
.PP
.I Cyr_expire
reads its configuration options out of the
.IR imapd.conf (5)
file unless specified otherwise by \fB-C\fR.
.SH OPTIONS
.TP
.BI \-C " config-file"
Read configuration options from \fIconfig-file\fR.
.TP
\fB\-E \fIdays\fR
Prune the duplicate database of entries older than \fIdays\fR.  This
value is only used for entries which do not have a corresponding
\fB/vendor/cmu/cyrus-imapd/expire\fR mailbox annotation.
.TP
.B \-v
Enable verbose output.
.SH FILES
.TP
.B /etc/imapd.conf
.SH SEE ALSO
.PP
\fBimapd.conf(5)\fR, \fBmaster(8)\fR


doc/news
This file is intended to keep track of user-visible changes to the
release, for the construction of the Athena release notes.

Changes from release 8.0 to release 8.1:

	* Athena version numbers are now dotted triplets, e.g. 8.1.0
	  or 8.1.12.  Paths which include all or part of the Athena
	  version number now take the form "pathname-version",
	  e.g. "srvd-8.1" or "dash.menus-8.1"

	* In release 8.0, running "xmkmf" with the default path would
	  generate a Makefile which looked for its include files and
	  libraries in /mit/x11, which could be one of several
	  lockers.  Now it is preferrable to use the vendor X
	  libraries, so this bit of glue is going away.  There will be
	  no /usr/athena/bin/xmkmf any more, nor any symlinks in
	  /usr/athena pointing to the x11 locker.

	  Users can still do something like "add -f x11r5" and then
	  "xmkmf", and things should work, if the x11r5 locker is
	  built for their platform.

	* Athena bitmaps have moved to /usr/athena/include/X11/bitmaps.
	  Previously, they were in /usr/athena/lib/X11/bitmaps before
	  because of the /mit/x11 glue.

	* The symlink /bin/csh -> /bin/athena/tcsh has gone away.  The
	  update to 8.0 will fix /bin/csh references in /etc/passwd
	  and /etc/passwd.local to refer to /bin/athena/tcsh instead.

	* emacs18 has gone away.

	* AFS includes and libraries are installed in
	  /usr/afsws/include and /usr/afsws/lib.

	* wcl includes and libraries are installed in /usr/athena.

	* The basic krb5 headers and libraries are installed in
	  /usr/athena.

	* emacs has been upgraded from 19.30 to 19.34.  There should
	  be no significant user-visible changes.

	* RCS has been upgraded from version 4 to version 5.7.

	* x3270 has been upgraded from version 1.2 to version 3.1.1.3.

	* sysinfo is now available under Irix.

	* /bin/athena/tcsh now has Hesiod support enabled, which means
	  that "~lockername" will work in a pathname in most cases.
	  (It won't work for users with local passwd information who
	  don't have their lockers attached.)

	* For improved ease of debugging, many binaries are installed
	  with symbol tables, and compiled without optimization.
	  Therefore, the contents of /bin/athena and /usr/athena/bin
	  may be somewhat larger, and the binaries may consume more
	  CPU cycles, although it's unlikely that anyone will notice
	  the speed difference.


third/cyrus-imapd/imap/cyr_expire.c
/* cyr_expire.c -- Program to expire deliver.db entries and messages
 *
 * Copyright (c) 1998-2003 Carnegie Mellon University.  All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 *
 * 1. Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer. 
 *
 * 2. Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in
 *    the documentation and/or other materials provided with the
 *    distribution.
 *
 * 3. The name "Carnegie Mellon University" must not be used to
 *    endorse or promote products derived from this software without
 *    prior written permission. For permission or any other legal
 *    details, please contact  
 *      Office of Technology Transfer
 *      Carnegie Mellon University
 *      5000 Forbes Avenue
 *      Pittsburgh, PA  15213-3890
 *      (412) 268-4387, fax: (412) 268-7395
 *      tech-transfer@andrew.cmu.edu
 *
 * 4. Redistributions of any form whatsoever must retain the following
 *    acknowledgment:
 *    "This product includes software developed by Computing Services
 *     at Carnegie Mellon University (http://www.cmu.edu/computing/)."
 *
 * CARNEGIE MELLON UNIVERSITY DISCLAIMS ALL WARRANTIES WITH REGARD TO
 * THIS SOFTWARE, INCLUDING ALL IMPLIED WARRANTIES OF MERCHANTABILITY
 * AND FITNESS, IN NO EVENT SHALL CARNEGIE MELLON UNIVERSITY BE LIABLE
 * FOR ANY SPECIAL, INDIRECT OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
 * WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN
 * AN ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING
 * OUT OF OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
 *
 * $Id: cyr_expire.c,v 1.1.1.1 2004-02-23 22:55:04 rbasch Exp $
 */

#include <config.h>

#ifdef HAVE_UNISTD_H
#include <unistd.h>
#endif
#include <stdlib.h>
#include <stdio.h>
#include <string.h>
#include <ctype.h>
#include <fcntl.h>
#include <sys/stat.h>
#include <syslog.h>
#include <com_err.h>
#include <errno.h>
#include <signal.h>

#include "annotate.h"
#include "cyrusdb.h"
#include "duplicate.h"
#include "exitcodes.h"
#include "global.h"
#include "hash.h"
#include "libcyr_cfg.h"
#include "mboxlist.h"
#include "util.h"
#include "xmalloc.h"

/* global state */
const int config_need_data = 0;

void usage(void)
{
    fprintf(stderr,
	    "cyr_expire [-C <altconfig>] -E <days> [-v]\n");
    exit(-1);
}

/*
 * mboxlist_findall() callback function to build a hash table of mailboxes
 * which need to have messages expired.
 */
int build_table(char *name, int matchlen, int maycreate __attribute__((unused)),
	       void *rock)
{
    struct hash_table *expire_table = (struct hash_table *) rock;
    char buf[MAX_MAILBOX_NAME+1] = "", *p;
    struct annotation_data attrib;
    int r, domainlen = 0;

    if (config_virtdomains && (p = strchr(name, '!')))
	domainlen = p - name + 1;

    strncpy(buf, name, matchlen);
    buf[matchlen] = '\0';

    /* since mailboxes inherit /vendor/cmu/cyrus-imapd/expire,
       we need to iterate all the way up to "" (server entry) */
    while (1) {
	r = annotatemore_lookup(buf, "/vendor/cmu/cyrus-imapd/expire", "",
				&attrib);

	if (r ||				/* error */
	    attrib.value ||			/* found an entry */
	    !buf[0] ||				/* done recursing */
	    !strcmp(buf+domainlen, "user")) {	/* server entry doesn't apply
						   to personal mailboxes */
	    break;
	}

	p = strrchr(buf, '.');			/* find parent mailbox */

	if (p && (p - buf > domainlen))		/* don't split subdomain */
	    *p = '\0';
	else if (!buf[domainlen])		/* server entry */
	    buf[0] = '\0';
	else					/* domain entry */
	    buf[domainlen] = '\0';
    }

    if (!r && attrib.value) {
	/* add mailbox to table */
	unsigned long days = strtoul(attrib.value, NULL, 10);
	time_t *expmark = (time_t *) xmalloc(sizeof(time_t));

	*expmark = days ? time(0) - (days * 60 * 60 * 24) : 0 /* never */ ;
	hash_insert(name, (void *) expmark, expire_table);
    }

    return r;
}

struct expire_rock {
    time_t expmark;
    unsigned long mailboxes;
    unsigned long messages;
    unsigned long deleted;
    int verbose;
};

/*
 * mailbox_expunge() callback to expunge expired articles.
 */
static int expunge_cb(struct mailbox *mailbox, void *rock, char *index)
{
    struct expire_rock *erock = (struct expire_rock *) rock;
    bit32 senttime = ntohl(*((bit32 *)(index+OFFSET_SENTDATE)));

    erock->messages++;

    if (senttime < erock->expmark) {
	erock->deleted++;
	return 1;
    }

    return 0;
}

static void do_expire(char *mboxname, void *data, void *rock)
{
    struct expire_rock *erock = (struct expire_rock *) rock;
    time_t *expmark = (time_t *) data;
    struct mailbox mailbox;
    int r;

    if (erock->verbose) {
	fprintf(stderr, "expiring messages in %s older than %ld days\n",
		mboxname, (time(0) - *expmark) / (60 * 60 * 24));
    }

    erock->mailboxes++;
    erock->expmark = *expmark;

    /* Open/lock header */
    r = mailbox_open_header(mboxname, 0, &mailbox);
    if (!r && mailbox.header_fd != -1) {
	(void) mailbox_lock_header(&mailbox);
	mailbox.header_lock_count = 1;
    }

    if (!r) r = chdir(mailbox.path);

    /* Attempt to open/lock index */
    if (!r) r = mailbox_open_index(&mailbox);
    if (!r) {
	(void) mailbox_lock_index(&mailbox);
	mailbox.index_lock_count = 1;
    }

    if (!r) mailbox_expunge(&mailbox, 1, expunge_cb, erock);
    mailbox_close(&mailbox);
}

int main(int argc, char *argv[])
{
    extern char *optarg;
    int opt, r = 0, days = 0;
    char *alt_config = NULL;
    char buf[100];
    struct hash_table expire_table;
    struct expire_rock erock;

    if (geteuid() == 0) fatal("must run as the Cyrus user", EC_USAGE);

    /* zero the expire_rock */
    memset(&erock, 0, sizeof(erock));

    while ((opt = getopt(argc, argv, "C:E:v")) != EOF) {
	switch (opt) {
	case 'C': /* alt config file */
	    alt_config = optarg;
	    break;

	case 'E':
	    if (days) usage();
	    days = atoi(optarg);
	    break;

	case 'v':
	    erock.verbose++;
	    break;
	
	default:
	    usage();
	    break;
	}
    }

    if (!days) usage();

    cyrus_init(alt_config, "cyr_expire");

    annotatemore_init(0, NULL, NULL);
    annotatemore_open(NULL);

    mboxlist_init(0);
    mboxlist_open(NULL);

    if (duplicate_init(NULL, 0) != 0) {
	fprintf(stderr, 
		"cyr_expire: unable to init duplicate delivery database\n");
	exit(1);
    }

    /* xxx better way to determine a size for this table? */
    construct_hash_table(&expire_table, 10000, 1);

    strlcpy(buf, "*", sizeof(buf));
    mboxlist_findall(NULL, buf, 1, 0, 0, &build_table, &expire_table);

    r = duplicate_prune(days, &expire_table);

    if (!r) {
	hash_enumerate(&expire_table, do_expire, &erock);

	syslog(LOG_NOTICE, "expunged %lu out of %lu messages from %lu mailboxes",
	       erock.deleted, erock.messages, erock.mailboxes);
	if (erock.verbose) {
	    fprintf(stderr, "\nexpunged %lu out of %lu messages from %lu mailboxes\n",
		    erock.deleted, erock.messages, erock.mailboxes);
	}
    }

    free_hash_table(&expire_table, free);

    mboxlist_close();
    mboxlist_done();
    annotatemore_close();
    annotatemore_done();
    duplicate_done();
    cyrus_done();

    exit(r);
}
